src/store/index-level.ts

```diff
@@ -216,46 +287,203 @@ export class IndexLevel {
+      // we also only include the cursor ONLY if it is relevant to the exact property in the 'lte' filter.
+      const lteCursor = cursor && this.extractValueFromKey(cursor) === this.encodeValue(rangeFilter.lte) ? cursor : undefined;
+      for (const item of await this.findExactMatches(tenant, propertyName, rangeFilter.lte, sortProperty, sortDirection, lteCursor, options)) {
+        lteMatches.push(item);
       }
     }
 
-    return matches;
+    // if the iterator is reversed and the results should be in Ascending order, we reverse the iterated matches.
+    // if the iterator is not reversed and the results should be in Descending order, we also reverse the iterated matches.
+    if ((iteratorOptions.reverse === true && sortDirection === SortOrder.Ascending) ||
+      (iteratorOptions.reverse !== true && sortDirection === SortOrder.Descending)) {
+      matches.reverse();
+    }
+
+    // the 'lteMatches' are already sorted, but depending on the sort we add them before or after the range matches.
+    if (sortDirection === SortOrder.Ascending) {
+      return [...matches, ...lteMatches];
+    } else {
+      return [...lteMatches, ...matches];
+    }
   }
 
-  private encodeValue(value: unknown): string {
-    switch (typeof value) {
-    case 'string':
-      // We can't just `JSON.stringify` as that'll affect the sort order of strings.
-      // For example, `'\x00'` becomes `'\\u0000'`.
-      return `"${value}"`;
-    case 'number':
-      return IndexLevel.encodeNumberValue(value);
-    default:
-      return String(value);
+  /**
+   * Construct a sortable index key to be used for each individual property of an indexed item.
+   * Although all of the properties are required to construct the key when inserting, we also use this function for creating a prefix.
+   *
+   * ex: sortProperty\u0000propertyName\u0000[propertyValue]\u0000[sortValue]\u0000[key]
+   *
+   * @param sortProperty the sorting property to construct the prefix of the key.
+   * @param propertyName the specific property name being indexed.
+   * @param propertyValue the optional specific property value being indexed.
+   * @param sortValue the optional value that determines the sort order in a lexicographical sorted way
+   * @param key the optional unique key of the item being indexed, this is used as a tiebreaker in case the other properties are the same.
+   * @returns a key to be used for sorted indexing within the LevelDB Index.
+   */
+  protected constructIndexedKey(sortProperty: string, propertyName: string, propertyValue?: string, sortValue?: string, key?: string): string {
+    const keyConstruction = [ sortProperty, propertyName ];
+    if (propertyValue !== undefined) {
+      keyConstruction.push(propertyValue);
+    }
+
+    if (sortValue !== undefined) {
+      keyConstruction.push(sortValue);
     }
+    if (key !== undefined) {
+      keyConstruction.push(key);
+    }
+
+    return this.join(...keyConstruction);
+  }
+
+  async getFilterCursors(
```

Comment:
This method is kinda vague to me upon encountering it, a few things I am not clear about:

1. Feels a summary description is useful given the complexity of the method.
1. Unsure why you'd need "multiple cursors".
1. Return type is.... unclear. Further glancing doesn't help, sometimes it returns `undefined`, sometimes it returns an empty object, and sometimes it returns an objects to Maps...

Would be useful to give some context in code here.

Reply:
Yeah... the use of `cursor` itself is overloaded (thus confusing for newcomer), sometimes it means the item ID, sometimes it means something else (key value as you mentioned).

We need to rename them to differentiate somehow. But looking pass all that, I am still trying to see how "sort value" is been taken advantage of, 'coz that's gotta be where the magic happens, especially in range queries. I'll keep reading, I am getting it bit by bit.... SLOOOOOOOWLY.